File: cs/asd_2_wquf_apps/src/unionfind/UnionFind.cs
/*
 * File: UnionFind.cs
 * Description: A base class implementing the QuickFind algorithm of 
 * dynamic connectivity of elements of the set.
 * Authors:
 *   - Ilya Tsivilskiy
 * Copyright: (c) 2023 Ilya Tsivilskiy
 * License: This file is licensed under the MIT License.
 */

using asd_2_collunionfind.src.utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace asd_2_collunionfind.src.unionfind
{
    internal class UnionFind : IUnionFind
    {
        protected int[] _id; // ID of a cluster that contains the connected elements
        protected bool _debug;

        /// <summary>
        /// Create a UnionFind data structure
        /// </summary>
        /// <param name="n">Total number of elements in collection</param>
        /// <param name="debug">Print out the connectivity info after each operation</param>
        public UnionFind(int n, bool debug = false)
        {
            // init an array of indices
            _id = new int[n];

            _debug = debug;

            // fill-up the array:
            // at first, each element's id equal its actual index,
            // so all elements form an individual own cluster
            for (int i = 0; i < n; i++)
            {
                _id[i] = i;
            }
        }

        // make the p-th element to belong to the cluster of the q-th element
        // union(4, 3) = connect 4 to the cluster of 3
        // NB: If a method in the base class is marked as "virtual",
        // it means that the method can be "overridden" in a derived class!
        virtual public void Union(int p, int q)
        {
            if (_debug)
            {
                Logger.WriteLine($"union({p}, {q}):");
                Logger.Write("before:\t");
                Print();
            }

            // no need to connect the elements,
            // if they are already connected
            if (_CallConnected(p, q))
            {
                return;
            }

            _CallUnion(p, q);

            if (_debug)
            {
                Logger.Write("after:\t");
                Print();
                Logger.WriteLine();
            }
        }

        // NB: If a method in the base class is marked as "virtual",
        // it means that the method can be "overridden" in a derived class!
        virtual public bool Connected(int p, int q)
        {
            // they belong to the same cluster
            bool result = _CallConnected(p, q);
            if (_debug)
            {
                Logger.WriteLine($"connected({p}, {q}) = {result}");
            }  
            return result;
        }

        /// <summary>
        /// The "core" of the "Union" method, 
        /// connects two elements
        /// </summary>
        /// <param name="p"></param>
        /// <param name="q"></param>
        virtual protected void _CallUnion(int p, int q)
        {
            int pCluster = _id[p]; // original p-th id = 4
            int qCluster = _id[q]; // original q-th id = 3
            // chain-style change of ids:
            // recursively connect the elements
            for (int i = 0; i < _id.Length; i++)
            {
                // if i-th element belongs to p-th cluster
                if (_id[i] == pCluster) // if id[i] == 4
                {
                    // connect it to the q-th cluster
                    _id[i] = qCluster; // id[i] = 3
                }
            }
        }

        /// <summary>
        /// The "core" of the "Connected" method, 
        /// checks whether two elememts are connected
        /// </summary>
        /// <param name="p"></param>
        /// <param name="q"></param>
        /// <returns></returns>
        virtual protected bool _CallConnected(int p, int q)
        {
            return _id[p] == _id[q];
        }

        public override string ToString()
        {
            StringBuilder str = new StringBuilder();
            foreach (int id in _id)
            {
                str.Append(id).Append(" ");
            }
            return str.ToString();
        }

        public void Print()
        {
            Logger.WriteLine(this.ToString());
        }
    }
}

File: java/asd_2_wquf_apps/src/unionfind/UnionFind.java
/*
 * File: UnionFind.java
 * Description: A base class implementing the QuickFind algorithm of
 * dynamic connectivity of elements of the set.
 * Authors:
 *   - Ilya Tsivilskiy
 * Copyright: (c) 2023 Ilya Tsivilskiy
 * License: This file is licensed under the MIT License.
 */

package unionfind;
import utils.Logger;

public class UnionFind implements IUnionFind {
    protected int[] id; // ID of a cluster that contains the connected elements
    protected boolean debug;

    /**
     * Create a UnionFind data structure
     * @param n     Total number of elements in collection
     * @param debug Print out the connectivity info after each operation
     */
    public UnionFind(int n, boolean debug) {
        // Initialize an array of indices
        id = new int[n];
        this.debug = debug;

        // Fill up the array:
        // At first, each element's id equals its actual index,
        // so all elements form an individual own cluster
        for (int i = 0; i < n; i++) {
            id[i] = i;
        }
    }

    @Override
    public void union(int p, int q) {
        if (debug) {
            Logger.writeLine(String.format("union(%d, %d):", p, q));
            Logger.write("before:\t");
            print();
        }

        // No need to connect the elements,
        // if they are already connected
        if (connected(p, q)) {
            return;
        }

        callUnion(p, q);

        if (debug) {
            Logger.write("after:\t");
            print();
            Logger.writeLine();
        }
    }

    @Override
    public boolean connected(int p, int q) {
        // They belong to the same cluster
        boolean result = callConnected(p, q);
        if (debug) {
            Logger.writeLine(String.format("connected(%d, %d) = %b", p, q, result));
        }
        return result;
    }

    /**
     * The "core" of the "union" method, connects two elements
     * @param p
     * @param q
     */
    protected void callUnion(int p, int q) {
        int pCluster = id[p]; // Original p-th id
        int qCluster = id[q]; // Original q-th id
        // Chain-style change of ids:
        // Recursively connect the elements
        for (int i = 0; i < id.length; i++) {
            // If i-th element belongs to p-th cluster
            if (id[i] == pCluster) {
                // Connect it to the q-th cluster
                id[i] = qCluster;
            }
        }
    }

    /**
     * The "core" of the "connected" method, checks whether two elements are connected
     * @param p
     * @param q
     * @return
     */
    protected boolean callConnected(int p, int q) {
        return id[p] == id[q];
    }

    @Override
    public String toString() {
        StringBuilder str = new StringBuilder();
        for (int anId : id) {
            str.append(anId).append(" ");
        }
        return str.toString();
    }

    @Override
    public void print() {
        Logger.writeLine(this.toString());
    }
}
